coap-gateway/service/clientObserveHandler.go

```diff
@@ -68,30 +68,35 @@ func CreateResourceContentToObserver(client *session, resourceChanged *events.Re
 }
 
 type resourceSubscription struct {
-	client   *session
-	token    coapMessage.Token
-	authCtx  *authorizationContext
-	deviceID string
-	href     string
-
-	seqNum uint32
-	sub    *subscription.Sub
-
-	mutex              sync.Mutex
+	client             *session
+	authCtx            *authorizationContext
+	sub                *subscription.Sub
+	deviceID           string
+	href               string
+	token              coapMessage.Token
 	version            uint64
+	mutex              sync.Mutex
+	seqNum             uint32
 	versionInitialized bool
 }
 
 func (s *resourceSubscription) cancelSubscription(code coapCodes.Code) {
+	x := struct {
+		s    *resourceSubscription
+		code coapCodes.Code
+	}{
+		s:    s,
+		code: code,
+	}
 	err := s.client.server.taskQueue.Submit(func() {
-		if _, err := s.client.cancelResourceSubscription(s.token.String()); err != nil {
-			s.client.Errorf("failed to cancel resource /%v%v subscription: %w", s.deviceID, s.href, err)
+		if _, err := x.s.client.cancelResourceSubscription(x.s.token.String()); err != nil {
+			x.s.client.Errorf("failed to cancel resource /%v%v subscription: %w", x.s.deviceID, x.s.href, err)
 		}
-		err := statusErrorf(code, "cannot observe resource /%v%v, device response: %v", s.deviceID, s.href, code)
-		resp := s.client.createErrorResponse(err, s.token)
-		defer s.client.ReleaseMessage(resp)
-		s.client.WriteMessage(resp)
-		s.client.logRequestResponse(nil, resp, err)
+		err := statusErrorf(code, "cannot observe resource /%v%v, device response: %v", x.s.deviceID, x.s.href, x.code)
```

Comment:
`statusErrorf(code, "cannot observe resource /%v%v, device response: %v", x.s.deviceID, x.s.href, x.code)`

Uses both `code` and `x.code`, I guess the intention is to use the `x` variable in the closure, so the `code` variable should be replaced.

Reply:
fixed